go/stats/pull_backend.go

```diff
@@ -0,0 +1,127 @@
+package stats
+
+import (
+	"sync"
+
+	log "github.com/golang/glog"
+)
+
+type PullBackend interface {
+	NewMultiCounter(*MultiCounters, string)
+	NewMultiCounterFunc(*MultiCountersFunc, string)
+	NewCounter(*Counters, string)
+	// Add more functions here that need to be registered with Prom
+}
+
+var (
+	be = struct {
+		sync.RWMutex
+		m map[string]PullBackend
+	}{m: make(map[string]PullBackend)}
+
+	pullMultiCounters = struct {
+		sync.RWMutex
```

Comment:
I feel like this can all be simplified a lot if we go with the approach I mentioned before where the pointer to the underlying counter / multicounter / multicounterfunc object is stored in the collector itself.

Then you wouldn't need all these separate locks -- you can have a single one for the registry of pull backends and pull metrics since the only coordination needs to be at initialization / registration time.